lib/Doctrine/ORM/PersistentCollection.php

```diff
@@ -692,23 +692,41 @@ public function unwrap()
     protected function doInitialize()
     {
         // Has NEW objects added through add(). Remember them.
-        $newObjects = [];
+        $newlyAddedDirtyObjects = [];
 
         if ($this->isDirty) {
-            $newObjects = $this->collection->toArray();
+            $newlyAddedDirtyObjects = $this->collection->toArray();
         }
 
         $this->collection->clear();
```

Comment:
I think the current solution is a decent one, as it keeps the responsibility of re-attaching the `NEW` records within the `PersistentCollection` implementation. Distributing it around multiple objects may be better for performance, but is actually a mess to track on the long term